extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -298,51 +299,40 @@ export class AmpA4A extends AMP.BaseElement {
             return null;
           }
 
-          // Alias for validCreativePromise's resolve function. This will be called
-          // if any of the fetched keys successfully validate the creative.
-          let resolveValidation;
-          // Promise that will resolve upon any successful validation.
-          const validCreativePromise = new Promise(resolve => {
-            resolveValidation = resolve;
-          });
-
-          // The following block defines an Array of Promises of ArrayBuffers,
-          // none of which will ever resolve. In the first instance of a
-          // successful creative verification, resolveValidation will be
-          // called with the creative as the argument.
-          const signatureVerificationPromises = this.keyInfoSetPromises_.map(
-              keyInfoSetPromise => {
-                return keyInfoSetPromise.then(keyInfoSet => {
-                  return Promise.all(keyInfoSet.map(keyInfoPromise => {
-                    return keyInfoPromise.then(keyInfo => {
-                      if (keyInfo) {
-                        return verifySignature(
-                            new Uint8Array(creativeParts.creative),
-                            creativeParts.signature,
-                            keyInfo)
-                        .then(isValid => {
-                          if (isValid) {
-                            resolveValidation(creativeParts.creative);
-                          }
-                        },
-                        err => {
-                          user().error('Amp Ad', err, this.element);
-                        });
+          // For each signing service, we have exactly one Promise,
+          // keyInfoSetPromise, that holds an Array of Promises of signing keys.
+          // So long as any one of these signing services can verify the
+          // signature, then the creative is valid AMP.
+          return any(this.keyInfoSetPromises_.map(keyInfoSetPromise => {
+            // Resolve Promise into Array of Promises of signing keys.
+            return keyInfoSetPromise.then(keyInfoSet => {
+              // As long as any one individual key of a particular signing
+              // service, keyInfoPromise, can verify the signature, then the
+              // creative is valid AMP.
+              return any(keyInfoSet.map(keyInfoPromise => {
+                // Resolve Promise into signing key.
+                return keyInfoPromise.then(keyInfo => {
+                  if (keyInfo) {
+                    // If the key exists, try verifying with it.
+                    return verifySignature(
+                        new Uint8Array(creativeParts.creative),
+                        creativeParts.signature,
+                        keyInfo)
+                    .then(isValid => {
+                      if (isValid) {
+                        return creativeParts.creative;
                       }
+                      return Promise.reject();
```

Comment:
I'm not in love with introducing rejection-as-flow-control in places where it didn't already exist, since rejecting is basically throwing an exception. I have an idea for an alternative way this can work, by building an actual async iterator type. (Sorry for the relatively delayed response; travel has that effect, unfortunately.)

The code could then look something like this:

```javascript
constructor() {
  /** @private {!AsyncIterator<!PublicKeyInfoDef>} */
  this.keyInfos_ = this.fetchAndImportKeys_();
}

/** @return {!AsyncIterator<!PublicKeyInfoDef>} */
fetchAndImportKeys_() {
  if (!isCryptoAvailable()) {
    return EMPTY_ASYNC_ITERATOR;
  }
  return SINGLETON_ASYNC_ITERATOR
      .flatMap(() => this.getSigningServiceNames())
      .map(serviceName => signingServerURLs[serviceName])
      .asyncMap(url => xhrFor(this.win).fetchJson(url, {mode: 'cors', method: 'GET'}))
      .flatMap(jwkSetObj => jwkSetObj.keys)
      .asyncMap(importPublicKey);
}

/**
 * @param {!Uint8Array} creative
 * @param {!Uint8Array} signature
 * @return {!Promise<boolean>}
 */
validateCreative_(creative, signature) {
  return this.keyInfos_
      .asyncMap(keyInfo => verifySignature(creative, signature, keyInfo))
      .some(isValid => isValid);
}
```

This is a slightly unfair comparison because this sample code doesn't have error handling, but I'm confident that that could be added as well without introducing disproportionate complexity.

@glevitzky, any thoughts? If you think this is a good idea, I could put it together on the flight home. Happy to talk or collaborate more if you want.